community/cypher/cypher-compiler-3.1/src/test/scala/org/neo4j/cypher/internal/compiler/v3_1/planner/logical/PlannerQueryRewriterTest.scala

```diff
@@ -0,0 +1,205 @@
+    is_rewritten_to(
+      """MATCH (a), (b)
+         RETURN DISTINCT a as a, b as b""")
+
+  assert_that(
+    """MATCH (a)
+        OPTIONAL MATCH (a)-[r]->(b)
+        RETURN DISTINCT a as a, r as r""").
+    is_not_rewritten()
+
+  assert_that(
+    """MATCH (a)
+       OPTIONAL MATCH (a)-[r]->(b)
+       RETURN count(distinct a) as x""").
+    is_rewritten_to(
+      """MATCH (a)
+         RETURN count(distinct a) as x""")
+
+  assert_that(
+    """MATCH (a)
+       OPTIONAL MATCH (a)-[r]->(b)-[r2]->(c) WHERE c.prop = b.prop
+       RETURN DISTINCT b as b""").
+    is_not_rewritten()
+
+  assert_that(
+    """OPTIONAL MATCH (f:DoesExist)
+       OPTIONAL MATCH (n:DoesNotExist)
+       RETURN collect(DISTINCT n.property) AS a, collect(DISTINCT f.property) AS b """).
+    is_not_rewritten()
+
+  assert_that(
+    """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b)-[r2:T2]->(c)
+          RETURN DISTINCT b as b""").
+    is_rewritten_to(
+      """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b) WHERE (b)-[:T2]->()
+          RETURN DISTINCT b as b""")
+
+  assert_that(
+    """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b)-[r2:T2]->(c) WHERE b:B
+          RETURN DISTINCT b as b""").
+    is_rewritten_to(
+      """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b) WHERE b:B and (b)-[:T2]->()
+          RETURN DISTINCT b as b""")
+
+  assert_that(
+    """MATCH (a)
+            OPTIONAL MATCH (a)-[r:T1]->(b)-[r2:T2]->(c) WHERE c.age <> 42
+            RETURN DISTINCT b as b""").
+    is_not_rewritten()
+
+  assert_that(
+    """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b)-[r2:T2]->(c) WHERE c:A:B and c.id = 42 and c.foo = 'apa'
+          RETURN DISTINCT b as b""").
+    is_rewritten_to(
+      """MATCH (a)
+          OPTIONAL MATCH (a)-[r:T1]->(b) WHERE (b)-[:T2]->(:A:B {foo: 'apa', id: 42})
+          RETURN DISTINCT b as b""")
+
+  case class rewriteTester(originalQuery: String) {
+    def is_rewritten_to(newQuery: String): Unit =
+      test(originalQuery) {
+        val expected = getUnionQueryFrom(newQuery.stripMargin)
+        val original = getUnionQueryFrom(originalQuery.stripMargin)
+
+        val result = original.endoRewrite(fixedPoint(rewriter))
+        assert(result === expected, "\nWas not rewritten correctly\n" + originalQuery)
+      }
+
+    def is_not_rewritten(): Unit = test(originalQuery) {
+      val query = getUnionQueryFrom(originalQuery.stripMargin)
+      val result = query.endoRewrite(fixedPoint(rewriter))
+      assert(result === query, "\nShould not have been rewritten\n" + originalQuery)
+    }
+  }
+
+  private def assert_that(originalQuery: String): rewriteTester = rewriteTester(originalQuery)
+
+  private def getUnionQueryFrom(query: String): UnionQuery = {
+    val ast = parseForRewriting(query).endoRewrite(flattenBooleanOperators)
+    val mkException = new SyntaxExceptionCreator(query, Some(DummyPosition(0)))
+    val semanticState = semanticChecker.check(query, ast, mkException)
```

Comment:
We may want to check that `semanticState` doesn't contain any errors here:

`sematicState.errors shouldBe empty`